### PR TITLE
mysql plugin - check if connection is stale before running query

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
1. Check if mysql plugin connection is stale. If so, then return stale connection exception which triggers a retry after establishing a new connection.
2.  Refactor execute method code to remove code duplication. 
3. Add a new test case to check if stale connection gets detected.

Fixes appsmithorg/appsmith #2032 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
1. By adding a new test case as part of build test. 
2. By executing a query against a stale connection. 
Screenshot 2020-12-07 at 4 39 03 PM

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java [hunk 1]
```diff
@@ -227,21 +227,31 @@ private boolean getIsSelectQuery(String query) {
...
+            return resultMono
+                    .flatMap(res -> {
+                        ActionExecutionResult result = new ActionExecutionResult();
+                        result.setBody(objectMapper.valueToTree(rowsList));
+                        result.setIsExecutionSuccess(true);
+                        System.out.println(Thread.currentThread().getName() + " In the MySqlPlugin, got action " +
+                                "execution result: " + result.toString());
+                        return Mono.just(result);
+                    })
+                    .onErrorResume(exception -> {
```
Comment (trishaanand) [inline app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java:243]: We shouldn't be doing on error resume. In the server side execute function, we handle the errors. Please remove this.

Comment (sumitsum) [reply app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java]: Done.

Comment (sumitsum) [reply app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java]: Done.

File: app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java [hunk 2]
```diff
@@ -186,12 +189,21 @@ private boolean getIsSelectQuery(String query) {
...
             }
 
-            boolean isSelectQuery = getIsSelectQuery(query);
+            boolean isSelectOrShowQuery = getIsSelectOrShowQuery(query);
             final List<Map<String, Object>> rowsList = new ArrayList<>(50);
-            Flux<Result> resultFlux = Flux.from(connection.createStatement(query).execute());
-
-            if(isSelectQuery) {
-                return resultFlux
+            Flux<Result> resultFlux = Flux.from(connection.validate(ValidationDepth.REMOTE))
```
Comment (trishaanand) [inline app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java:194]: connection.validate seems to be returning a Mono which we are casting to Flux. Would that be a problem?

Comment (trishaanand) [reply app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java:194]:
```suggestion
            Flux<Result> resultFlux = Mono.from(connection.validate(ValidationDepth.REMOTE))
                    .flatMapMany(isValid -> {
                        if (isValid) {
                            return connection.createStatement(query).execute();
                        } else {
                            return Flux.error(new StaleConnectionException());
                        }
                    });
```

Comment (sumitsum) [reply app/server/appsmith-plugins/mysqlPlugin/src/main/java/com/external/plugins/MySqlPlugin.java]: Done.

**Review by trishaanand — DISMISSED** [review]

File: app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java [hunk 3]
```diff
@@ -209,6 +209,25 @@ public void testExecute() {
                 .verifyComplete();
     }
 
+    @Test
+    public void testStaleConnectionCheck() {
+        ActionConfiguration actionConfiguration = new ActionConfiguration();
+        actionConfiguration.setBody("show databases");
+        Connection connection = pluginExecutor.datasourceCreate(dsConfig).block();
+
+        try {
```
Comment (trishaanand) [inline app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java:218]: Please rewrite the test case :
StepVerifier
                .create(resultFlux)
                .expectErrorMatches(throwable -> throwable instanceof StaleConnectionException)
                .verify();

Comment (sumitsum) [reply app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java]: Done.

File: app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java [hunk 4]
```diff
@@ -209,6 +210,21 @@ public void testExecute() {
...
 
+    @Test
+    public void testStaleConnectionCheck() {
+        ActionConfiguration actionConfiguration = new ActionConfiguration();
+        actionConfiguration.setBody("show databases");
+        Connection connection = pluginExecutor.datasourceCreate(dsConfig).block();
+
+        Flux<ActionExecutionResult> resultFlux = Mono.from(connection.close())
+                .thenMany(pluginExecutor.execute(connection, dsConfig, actionConfiguration))
+                .map(res -> res);
```
Comment (trishaanand) [inline app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java:221]: Why is this line required? Can we remove this?

Comment (sumitsum) [reply app/server/appsmith-plugins/mysqlPlugin/src/test/java/com/external/plugins/MySqlPluginTest.java]: Done.

**Review by trishaanand — APPROVED** [review]
